fairseq2n/src/fairseq2n/data/composite_data_source.cc

```diff
@@ -0,0 +1,114 @@
+// Copyright (c) Meta Platforms, Inc. and affiliates.
+// All rights reserved.
+//
+// This source code is licensed under the BSD-style license found in the
+// LICENSE file in the root directory of this source tree.
+
+#include "fairseq2n/data/composite_data_source.h"
+
+namespace fairseq2n::detail {
+
+composite_data_source::composite_data_source(
+    std::vector<data_pipeline> &&pipelines,
+    index_generator_fn &&index_gen_fn,
+    bool stop_at_shortest)
+  : pipelines_(std::move(pipelines)),
+    next_index_gen_(std::move(index_gen_fn)),
```

Comment:
Nit: for this assignment and the next one for `stop_at_shortest_`, ideally use brace initialization (i.e. `next_index_gen_{std::move(..)}, stop_at_shorted_{stop_at_shortest}` to follow the convention in the rest of the code base.